pwiz_tools/Skyline/Executables/SkylineBatch/SkylineBatch/MainSettings.cs

```diff
@@ -0,0 +1,162 @@
+
+
+        public MainSettings(string templateFilePath, string analysisFolderPath, string dataFolderPath,
+            string replicateNamingPattern)
+        {
+            TemplateFilePath = templateFilePath;
+            AnalysisFolderPath = analysisFolderPath;
+            DataFolderPath = dataFolderPath;
+            ReplicateNamingPattern = replicateNamingPattern ?? "";
+            Validate();
+        }
+
+        public readonly string TemplateFilePath;
+
+        public readonly string AnalysisFolderPath;
+
+        public readonly string DataFolderPath;
+
+        public readonly string ReplicateNamingPattern;
+
+        public string GetNewTemplatePath()
+        {
+            return AnalysisFolderPath + "\\" + Path.GetFileName(TemplateFilePath);
+        }
+
+        public override string ToString()
+        {
+            var sb = new StringBuilder();
+            sb.Append("Template file: ").AppendLine(TemplateFilePath);
+            sb.Append("Analysis folder: ").AppendLine(AnalysisFolderPath);
+            sb.Append("Data folder: ").AppendLine(DataFolderPath);
+            sb.Append("Replicate naming pattern: ").AppendLine(ReplicateNamingPattern);
+            return sb.ToString();
+        }
+
+        public void Validate()
+        {
+            CheckIfEmptyPath(TemplateFilePath, "Skyline file");
+            if (!File.Exists(TemplateFilePath))
+            {
+                throw new ArgumentException(string.Format(Resources.Template_file_does_not_exist, TemplateFilePath));
+            }
+            CheckIfEmptyPath(AnalysisFolderPath, "analysis folder");
+            var analysisFolderDirectory = Path.GetDirectoryName(AnalysisFolderPath);
+            if (!Directory.Exists(analysisFolderDirectory))
+            {
+                throw new ArgumentException(string.Format(Resources.Analysis_folder_does_not_exist, analysisFolderDirectory));
+            }
+            CheckIfEmptyPath(DataFolderPath, "data folder");
+            if (!Directory.Exists(DataFolderPath))
+            {
+                throw new ArgumentException(string.Format(Resources.Data_folder_does_not_exist, DataFolderPath));
+            }
+            // create analysis folder if doesn't exist
+            if (!Directory.Exists(AnalysisFolderPath))
+            {
+                Directory.CreateDirectory(AnalysisFolderPath);
+            }
+        }
+
+        public void CheckIfEmptyPath(string input, string name)
+        {
+            if (string.IsNullOrWhiteSpace(input))
+            {
+                throw new ArgumentException(string.Format(Resources.Specify_path_to, name));
+            }
+        }
+
+        
+
+        
+
+
+        #region Read/Write XML
+
+        private enum Attr
+        {
+            TemplateFilePath,
+            AnalysisFolderPath,
+            DataFolderPath,
+            ReplicateNamingPattern,
+        };
+
+        public static MainSettings ReadXml(XmlReader reader)

```

Comment:
Did you know that you effectively get an implementation of "ReadXml" and "WriteXml" for free just by adding the [XmlRoot] attribute at the top of the class. You only need to implement IXmlSerializable if you need to have your own custom implementation, which we do in Skyline because we need to be able to handle older versions of the way that we used to write out the XML. If you don't care about backwards or forwards compatibility you can usually skip implementing ReadXml and WriteXml.